### ISSUE TITLE
Standard API for passing a series or value as arguments to a GeoSeries transform function

### ISSUE THREAD

**Issue by stuartlynn** [kylebarron/geopolars#40]
Currently, the arguments for functions such as scale, skew, rotate etc, are passed as a single value that is applied to the entire GeoSeries. Often, it will be more useful to optionally pass in a secondary series of the same length as the GeoSeries for these parameters and lookup the value to be scaled, skewed etc from that series.

A couple of ways we could do this

1. Create a second version of each function for example `scale_by_series( x_fact: &Series, y_fact: &Series)`   etc  that takes series and applies the transform. The downside of this is if you wanted to have a fixed x_factor scaling but a variable y_factor, it would require you create a series for x_fact that is all just the same value 

2. Create an enum something like the following and pass this in for each argument of the transform. This is more flexible but more verbose.

```rust
enum ValOrSeries<T>{
   Val(T), 
   Series(&Series)
}
```

3. Something else I haven't thought of yet

Eventually, it would be good to have these defined on the DataFrame in such a way that you could specify a column name, series or value as each of the input parameters

**Comment by stuartlynn** [comment]
@kylebarron do you have a preference here? I am leaning towards 2 myself as I think it allows for more flexibility

**Comment by kylebarron** [comment]
> Often, it will be more useful to optionally pass in a secondary series of the same length as the GeoSeries for these parameters and lookup the value to be scaled, skewed etc from that series

So essentially numpy-like broadcasting? It looks like Rust doesn't support overloading. I haven't thought about this before but I'm guessing that 2 would indeed be simpler for when you have multiple arguments.

**Comment by stuartlynn** [comment]
yeah like numpy basically. I can also take a look to see if there is an equivalent behavior in Polars and go with something similar

**Comment by stuartlynn** [comment]
Woke up this morning with an idea for this. Basically we use generics + the Into trait to kind of emulate function overloading 

```rust

use num::FromPrimitive;

pub struct Series{
    pub len: usize
}

pub enum PolarsParameter<T>{
   Val(T),
   Series(Series)
}

impl <T> From<T> for PolarsParameter<T>
    where T : FromPrimitive 
    {
    fn from(val:T) -> PolarsParameter<T>{
        PolarsParameter::Val(val) 
    }
}

impl <T> From<Series> for PolarsParameter<T>{
    fn from(val:Series) -> PolarsParameter<T>{
        return PolarsParameter::Series(val) 
    }
}

pub fn test_fn(x:impl Into<PolarsParameter<f64>>, y: impl Into<PolarsParameter<i32>>){
    match x.into(){
        PolarsParameter::Val(val)=> println!("X is a single value {}",val),
        PolarsParameter::Series(s) => println!("X is a series of length {}",s.len)
    };

    match y.into(){
        PolarsParameter::Val(val)=> println!("Y is a single value: {}",val),
        PolarsParameter::Series(s) => println!("Y is a series {}",s.len)
    };

}

fn main(){
    test_fn(20.0, 1);
    let x_series = Series{ len: 10};
    let y_series = Series{ len: 20};
    test_fn(2.0, y_series);
    test_fn(x_series, 3);
}
```
Where series here is a placeholder for the polars Series. In practice we probably want to make this a reference &Series which might get us in to trouble with lifetimes but I am sure we can figure it out.

We might even be able to change the From to a TryFrom and check if the series is of the correct type for the required variable type. 

We can add a From Impl for string or other types of parameter, for example Point  we would want to be able to pass to a function. 

You can play with this example here https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b668009700263240451bc29bc3d76b15
